pkgs/build-support/neowrap.nix

```diff
@@ -0,0 +1,294 @@
+      key:
+      value:
+      (let
+        sep = encyclopedia.separators.${key} or ":"; # default separator used for most wrappings
+      in
+        # To filter out envInfo changed via the `linkByEnv` argument
+        if builtins.isString (builtins.elemAt value 0) then
+          if builtins.elem key encyclopedia.singleValue then
+            if (builtins.length value) > 1 then
+              abort "neowrap.nix: There is more then 1 derivation in all inputs of: ${builtins.toJSON allPkgs} that set ${key} to a value via a passthru. This env key, is defined in my encyclopedia as a single value key. Hence I don't know what value to use in wrapping."
+            else
+              # Should this be "--set" ?
+              "--set-default ${key} ${builtins.elemAt value 0}"
+          else
+            "--prefix ${key} ${sep} ${builtins.concatStringsSep sep value}"
+        else
+          # duplicates are removed by lib.lists.flatten at the beginning
+          if (builtins.elemAt value 0).linkType == "normal" then
+            "--prefix ${key} ${sep} ${(builtins.elemAt value 0).linkTo}"
+          else # (builtins.elemAt value 0).linkType == "pkg" then
+            if builtins.hasAttr key wrapOut || builtins.hasAttr key encyclopedia.linkPaths then
+              "--prefix ${key} ${sep} ${wrapOut.${key} or encyclopedia.linkPaths.${key}}"
+            else
+              abort "neowrap.nix: When requested to link packages of env var ${key}, I couldn't decide where to link them _to_, either set wrapOut.${key} to e.g $out/${key}, and that directory will be added to the values of ${key}, or add an entry to the encyclopedia.linkPaths attribute set"
+      )
+    )
+      # Before calculating makeWrapperArgs, we need to add values to env vars
+      # according to the `wrapOut` argument:
+      (lib.attrsets.mapAttrs (
+        name:
+        values:
+        if builtins.hasAttr name wrapOut && builtins.isList values then
+          if builtins.elem wrapOut.${name} values then
+            values
+          else
+            values ++ [wrapOut.${name}]
+        else
+          values
+      ) envInfo)
+    );
+    makeWrapperArgs_ = builtins.trace "makeWrapperArgs is ${builtins.toJSON makeWrapperArgs}" makeWrapperArgs;
+    linkCmds = lib.lists.flatten (lib.attrsets.mapAttrsToList (
+      key:
+      values:
+      (let
+        sep = encyclopedia.separators.${key} or ":"; # default separator used for most wrappings
+      in
+        if builtins.isAttrs (builtins.elemAt values 0) then
+          (map (
+            v:
+            "mkdir -p ${v.linkTo} && lndir -silent ${v.linkFrom} ${v.linkTo}"
+          ) values)
+        else
+          # removed by lib.lists.flatten at the beginning
+          []
+      )
+    )
+      envInfo
+    );
+    # Useful for debugging, not evaluated if not used.
+    linkCmds_ = builtins.trace "linkCmds is ${builtins.toJSON linkCmds}" linkCmds;
+  in
+  symlinkJoin {
+    name = "runtime-env";
```

Comment:
Perhaps this should inherit the name of the parent derivation.

Reply:
Not always there's a single parent derivation, but certainly for most purposes, this would be nicer.